src/data/nat/factorization.lean

```diff
@@ -196,4 +199,65 @@ begin
   { rintro ⟨c, rfl⟩, rw factorization_mul hd (right_ne_zero_of_mul hn), simp },
 end
 
+lemma prime_pow_dvd_iff_factorization_ge (p k n : ℕ) (pp : prime p) (hn : n ≠ 0) :
+  p ^ k ∣ n ↔ k ≤ n.factorization p :=
+begin
+  rw [←(factorization_le_iff_dvd (ne_of_gt (pow_pos (prime.pos pp) k)) hn),
+    prime.factorization_pow pp, single_le_iff],
+end
+
+lemma exists_factorization_lt_of_lt {a b : ℕ} (ha : a ≠ 0) (hb : b ≠ 0) (hab : a < b) :
+  ∃ p : ℕ, a.factorization p < b.factorization p :=
+begin
+  contrapose! hab,
+  rw [←finsupp.le_def, factorization_le_iff_dvd hb ha] at hab,
+  exact le_of_dvd ha.bot_lt hab,
+end
+
+@[simp]
+lemma div_factorization_eq_tsub_of_dvd {d n : ℕ} (hd : d ≠ 0) (hn : n ≠ 0) (h : d ∣ n) :
+  (n/d).factorization = n.factorization - d.factorization :=
+begin
+  cases dvd_iff_exists_eq_mul_left.mp h with c hc,
+  have hc_pos : c ≠ 0, { subst hc, exact left_ne_zero_of_mul hn },
+  rw [hc, nat.mul_div_cancel c hd.bot_lt, factorization_mul hc_pos hd, add_tsub_cancel_right],
+end
+
+lemma dvd_iff_div_factorization_eq_tsub (d n : ℕ) (hd : d ≠ 0) (hn : n ≠ 0) (hdn : d ≤ n) :
+  d ∣ n ↔ (n/d).factorization = n.factorization - d.factorization :=
+begin
+  split,
+  { exact div_factorization_eq_tsub_of_dvd hd hn },
+  { rcases eq_or_lt_of_le hdn with rfl | hd_lt_n, { simp },
+    have h1 : n / d ≠ 0 := λ H, nat.lt_asymm hd_lt_n ((nat.div_eq_zero_iff hd.bot_lt).mp H),
+    intros h,
+    rw dvd_iff_le_div_mul n d,
+    by_contra h2,
+    cases (exists_factorization_lt_of_lt (mul_ne_zero h1 hd) hn (not_le.mp h2)) with p hp,
+    rwa [factorization_mul h1 hd, add_apply, ←lt_tsub_iff_right, h, tsub_apply,
+      lt_self_iff_false] at hp },
+end
+
+lemma pow_factorization_dvd (p d : ℕ) : p ^ (d.factorization) p ∣ d :=
+begin
+  rcases em (d = 0) with rfl | hd, { simp },
```

Comment:
The canonical way of writing this would be
```
rcases eq_or_ne d 0 with rfl | hd, { simp },
```

Reply:
Thanks!